### ISSUE TITLE
Mongoose accepts manually supplied ObjectID field although it's created automatically

### ISSUE THREAD

**Issue by butlimous** [Automattic/mongoose#8746]
**Do you want to request a *feature* or report a *bug*?**
Report a bug or make sure it's by design

**What is the current behavior?**
If I have the following simple schema:

`{
     name: { type: String },
    age: { type: Number }
  }`

And I create a new document for this model like this:

` const record = {
      _id: '5e35517cc894c90327a34bae',
     name: 'Bob',
     age: 30
 }
Model.create(record);`

The record gets created successfully but with the `_id` field that I manually supplied in the request! It does the same behavior for subdocuments (array of objects)

**What is the expected behavior?**

As long as I haven't specified a different format for the `_id` field in the schema like `_id: {type:number}` for example, the database should act like SQL ones with autoincrement and just ignore the field in the request to create the document. Otherwise, I would have to manually remove the field from the request for each and every field that has the automatically generated ObjectID field. I get the request fields through an API so I have to properly secure this.

To summarize, the `_id` field should automatically get created and ignore any manually supplied `_id`  unless explicitly stated in the schema. This is actually the behavior for any non existing field (it gets ignored). It should be the same for the `_id` field or am I missing something?

**Comment by AbdelrahmanHafez** [comment]
This is by design, mongoose (and MongoDB) allow you to set the `_id` manually, in order to secure your API, I would recommend validating user input using [joi](https://www.npmjs.com/package/@hapi/joi) which disallows the user from sending any other fields than the ones you're specifying.

If you're allowing a user to enter data without validation, them inserting an `_id` should be the least of your concerns.

Please let me know if you need any further assistance with this issue.

**Comment by vkarpov15** [comment]
Allowing explicitly setting an `_id` on a new document is by design. Why do you have to "properly secure this"? Is there some sort of security issue?

**Comment by butlimous** [comment]
Of course, it's a major security issue! 
For my use case, I receive the info used to create the document from user input so a hacker could screw the database by manually supplying the _ids

It's easy to validate this but it doesn't make sense to be able to modify something that is automatically generated. 

I assumed that it works like SQL auto increment and would just ignore the supplied id. It only makes sense to accept the field when I explicitly add it to the schema.

**Comment by AbdelrahmanHafez** [comment]
This behavior is by design and is supported since day 1, limiting the package users to not be able to enter an `_id` would be a huge breaking change, that would be a huge effort for users to switch to that new behavior.

Let alone that this design is also provided by [MongoDB and its drivers](https://docs.mongodb.com/manual/reference/method/db.collection.insertOne/#id-field).

It's probably a better approach to validate, and as someone who learned that lesson the difficult way, use @hapijs/joi, it's pretty much the best validation library out there.

I'll close the issue, please let me know if you need any further assistance.

**Comment by vkarpov15** [comment]
How could a hacker screw the database by manually supplying `_id`? I might be missing something here.

I strongly disagree about Joi. In general, you should be able to use Mongoose to insert or update user-modified data. But I'd recommend using [archetype](https://www.npmjs.com/package/archetype) if you need a separate casting/validation library.

**Comment by AbdelrahmanHafez** [comment]
Well, I wouldn't let users manually enter an `_id`, maybe he's counting on the `_id` to get the creation timestamp.

I am not recommending that people should use Joi on top of mongoose or as an alternative to mongoose validation. What I am recommending is for people to use Joi to validate user input on the HTTP layer of things, like validating `req.body`.

Do you disagree with that? If so, I'd love to learn why.

**Comment by butlimous** [comment]
@vkarpov15 `_id` is the main method for manipulating the data in the database so changing it would screw tasks like sorting the data and of course I need to be aware of anything that gets done inside the DB

@AbdelrahmanHafez I'm creating an internal express API to exchange data between frontend and backend so I need to properly secure this. The API is for internal usage only but the bad guys could try to use it externally and mess things up

**Comment by vkarpov15** [comment]
@AbdelrahmanHafez yeah it looks like OP is concerned with sorting by `_id`. Which I don't recommend doing anyway, you're better off creating an index on `createdAt` because you typically want to sort documents in reverse order (newest first), and the `_id` index is always `_id: 1`.

@butlimous I'd challenge you to come up with a case where this behavior is actually a problem. "Bad guys could mess things up" doesn't really qualify as a security issue report.
